setup-deploy-keys/src/clean_unused_checkouts.rs

```diff
@@ -0,0 +1,235 @@
+use clap::Parser;
+use std::fs;
+use std::io;
+use std::path::{Path, PathBuf};
+use std::time::{Duration, SystemTime};
+use walkdir::WalkDir;
+
+/// Clean up unused projects
+///
+/// This CLI finds all projects that users have checked out on the dev-desktops and deletes
+/// temporary files if the project has not been modified in a certain number of days.
+///
+/// Specifically, the CLI will look for checkouts of `rust-lang/rust` and delete the `build`
+/// directory. And it will find unused crates and delete the `target` directory.
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Cli {
+    /// Only print directories and their size, do not delete
+    #[arg(long)]
+    dry_run: bool,
+
+    /// The root directory to search for projects
+    #[arg(short, long = "root-directory", default_value = "/home")]
+    root_directory: PathBuf,
+
+    /// The maximum age of a project in days
+    ///
+    /// The CLI will only clean projects that have not been updated in the last `max-age` days.
+    #[arg(short, long = "max-age", default_value_t = 60)]
+    max_age: u32,
+}
+
+fn find_cache_dirs(home: &Path) -> io::Result<Vec<PathBuf>> {
+    // Use WalkDir to perform a safe recursive traversal. By default WalkDir does
+    // not follow symlinks which prevents accidental symlink loops.
+    let mut result = Vec::new();
+
+    for entry in WalkDir::new(home).follow_links(false) {
+        let entry = match entry {
+            Ok(e) => e,
+            Err(_) => continue, // skip entries we can't read
+        };
+
+        let path = entry.path();
+        if !entry.file_type().is_dir() {
+            continue;
+        }
+
+        // We're interested in artifact dirs named `build` (python) or
+        // `target` (Rust). When we find one, check the parent directory for the
+        // expected marker files (`x.py` for python projects, `Cargo.toml` for
+        // Rust) before including the artifact directory.
+        if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
+            match name {
+                "build" => {
+                    if path
+                        .parent()
+                        .map(|p| p.join("x.py").is_file())
+                        .unwrap_or(false)
+                    {
+                        result.push(path.to_path_buf());
+                    }
+                }
+                "target" => {
+                    if path
+                        .parent()
+                        .map(|p| p.join("Cargo.toml").is_file())
+                        .unwrap_or(false)
+                    {
+                        result.push(path.to_path_buf());
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+
+    result.sort();
+    result.dedup();
```

Comment:
Nitpick: since we don't want duplicates and we don't care about ordering, we could use a hashset instead of a vec